### DISCUSSION TITLE
Issue with `generateBuildId`–– assets not identical between apps (using load balancer)

### DISCUSSION THREAD

**Discussion by andrewmartin** [vercel/next.js#22597] [Help]
Hi there,

I've had a LOT of success with this great framework but am hitting my head against the wall on this one.  I'm deploying a Next.js app on a server using a straightforward and simple `next build` command. I'm using a common build ID and I'm not getting the same output between builds, leading to issues on ELB when I have two separate app servers serving the same code.

I did find this issue, which seemed helpful, but curious if there's any workaround that doesn't involve me upgrading to webpack 5?
https://github.com/vercel/next.js/issues/20352#issue-771815605

![Image 2021-02-27 at 6 41 37 PM](https://user-images.githubusercontent.com/273842/109406536-c2e24f80-792e-11eb-8ba4-71b04cbb5567.jpg)

My `next.config.js`:

```js
const withGraphql = require('next-plugin-graphql')
const nextBuildId = require('next-build-id')

module.exports = withGraphql({
  generateBuildId: () => nextBuildId({ dir: __dirname }),
})
```

Happy to share any more info you might need, and thanks in advance for your help.

**Comment by andrewmartin** [comment]
Always think a picture is worth a thousand words, here's another screenshot:

![Image 2021-02-27 at 6 41 01 PM](https://user-images.githubusercontent.com/273842/109406568-f7eea200-792e-11eb-93b0-c10a6def5f74.jpg)

**Comment by andrewmartin** [comment]
I should clarify; the CSS file does not show up, the JS file does.

**Comment by andrewmartin** [comment]
I think I found the issue! but would love some advice.

I have two apps on two different servers using a load balancer. I'm provisioning them using ansible, so they're both built from the same git commit at the same time. However, I need _all_ of the static assets to completely and totally match up between both of the servers, of course. Basically, I need both servers to be provisioned and built using _exactly the same code_.

What I've done in the past is this:

1. Generate a file with some `git rev-parse` info
2. Make `next.config.js` reference this file in the `generateBuildId` function
3. Enjoy.

But, for some reason, this process isn't working anymore. Any thoughts or better patterns?

`HEAD-CLIENT.js`
```
module.exports = 'f619bfa trying next-build-id'
```

```
module.exports = withGraphql({
  generateBuildId: async () => {
    // eslint-disable-next-line global-require
    const BUILDID = require('../HEAD-CLIENT.js')
    // eslint-disable-next-line no-console
    console.info(`building from: ${BUILDID}`)

    return BUILDID
  },
})
```

Also, just to confirm, on both of my instances, I have different output from the builds, but the `BUILD_ID` is identical in both:


Server one (redacted most of the build-manifest, you get the idea):
```
[ubuntu@ec2-3-(REDACTED).us-west-1 .next]$ cat BUILD_ID
2ab99be reverting back build id

[ubuntu@ec2-3-(REDACTED).us-west-1 .next]$ cat build-manifest.json
    ...(REDACTED)...
    "/user/settings": [
      "static/chunks/main-06621f95abea4f74cc17.js",
      "static/chunks/webpack-9fe5efcb57f8ee56b776.js",
      "static/chunks/framework.0ba53a8de0b1a5f4cfd1.js",
      "static/chunks/commons.d657bb330032b1d92f83.js",
      "static/chunks/fd2972fd578cf174ab8096060d1d2e554cfe07ec.4521014ec666dee5f411.js",
      "static/chunks/45742203f4002a6463f47b0d0f09aaa6deff5b24.e1609703977fe7a5803b.js",
      "static/chunks/fc5ff07dfb42330bf047278fba6ea03a30a88d2a.6a9b312f5c117c49aaac.js",
      "static/css/058a6a0f060c52647176.css", // <- DIFFERENT!
      "static/chunks/pages/user/settings-5a8b9806507b826b0aa1.js"
    ]
  },
  "ampFirstPages": []
}
```

Server two:
```
[ubuntu@ec2-54-(REDACTED).us-west-1 .next]$ cat BUILD_ID
2ab99be reverting back build id

[ubuntu@ec2-54-(REDACTED).us-west-1 .next]$ cat server-manifest.json
    ...(REDACTED)...
    "/user/settings": [
      "static/chunks/main-06621f95abea4f74cc17.js",
      "static/chunks/webpack-8849b16fd73e0b82bf35.js",
      "static/chunks/framework.0ba53a8de0b1a5f4cfd1.js",
      "static/chunks/commons.d657bb330032b1d92f83.js",
      "static/chunks/fd2972fd578cf174ab8096060d1d2e554cfe07ec.4521014ec666dee5f411.js",
      "static/chunks/45742203f4002a6463f47b0d0f09aaa6deff5b24.e1609703977fe7a5803b.js",
      "static/chunks/fc5ff07dfb42330bf047278fba6ea03a30a88d2a.6a9b312f5c117c49aaac.js",
      "static/css/15d11cc7f40d3e5dacd9.css", // <- DIFFERENT!
      "static/chunks/pages/user/settings-d8ff1696073abd923208.js"
    ]
  },
  "ampFirstPages": []
}
```

**Reply by cullylarson** [reply]
Are the pathnames of the project on all of your servers the same? For example, is the project on one server in `/home/user1/my-project` and on another server it's in `/home/user1/Projects/my-project`? Different paths will produce different asset names.

**Reply by andrewmartin** [reply]
@cullylarson that may be it. Is there any options I can pass to override this setting that you know of?

**Reply by cullylarson** [reply]
@andrewmartin Not that I know of. I reported it in the issue you commented on earlier today (https://github.com/vercel/next.js/issues/20352#issuecomment-787514914). Someone commented that it would potentially be fixed in webpack 5. If it hasn't been, submitting an issue to the webpack repo might be a good next step.

**Reply by andrewmartin** [reply]
Thanks for your replies. I do know that I have different paths for each of the deployments. I'm going to see about adjusting that.
